org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ProcessTaskManager.java

```diff
@@ -46,159 +51,131 @@ public class ProcessTaskManager {
 		return t;
 	});
 
-public ProcessTaskManager(Compiler compiler, int startingIndex) {
-	this.compiler = compiler;
-	this.startingIndex = startingIndex;
+	public ProcessTaskManager(Compiler compiler, int startingIndex) {
+		this.compiler = compiler;
+		this.startingIndex = startingIndex;
 
-	this.currentIndex = 0;
-	this.availableIndex = 0;
-	this.size = PROCESSED_QUEUE_SIZE;
-	this.sleepCount = 0; // 0 is no one, +1 is the processing thread & -1 is the writing/main thread
-	this.units = new CompilationUnitDeclaration[this.size];
+		this.units = new ArrayBlockingQueue<>(PROCESSED_QUEUE_SIZE);
 
-	synchronized (this) {
-		this.processingTask = executor.submit(this::compile);
+		this.processingTask = executor.submit(this::processing);
 	}
-}
 
 // add unit to the queue - wait if no space is available
-private synchronized void addNextUnit(CompilationUnitDeclaration newElement) {
-	while (this.units[this.availableIndex] != null) {
-		//System.out.print('a');
-		//if (this.sleepCount < 0) throw new IllegalStateException(Integer.valueOf(this.sleepCount).toString());
-		this.sleepCount = 1;
+	private void addNextUnit(Object newElement) {
 		try {
-			wait(250);
-		} catch (InterruptedException ignore) {
-			// ignore
+			this.units.put(newElement);
+		} catch (InterruptedException interrupt) {
+			throw new RuntimeException(interrupt);
 		}
-		this.sleepCount = 0;
 	}
 
-	this.units[this.availableIndex++] = newElement;
-	if (this.availableIndex >= this.size)
-		this.availableIndex = 0;
-	if (this.sleepCount <= -1)
-		notify(); // wake up writing thread to accept next unit - could be the last one - must avoid deadlock
-}
+	private Object lastSignal = null;
 
-public CompilationUnitDeclaration removeNextUnit() throws Error {
-	CompilationUnitDeclaration next = null;
-	boolean yield = false;
-	synchronized (this) {
-		next = this.units[this.currentIndex];
-		if (next == null || this.caughtException != null) {
-			do {
-				if (this.processingTask == null) {
-					if (this.caughtException != null) {
-						// rethrow the caught exception from the processingThread in the main compiler thread
-						if (this.caughtException instanceof Error)
-							throw (Error) this.caughtException;
-						throw (RuntimeException) this.caughtException;
-					}
-					return null;
+	/** returns null when no more elements can be expected **/
```

Comment:
Can we return empty collection instead? It is always more convenient on the caller side to check for empty and not for null/not null

Reply:
Currently the call blocks until some elements are returned or end reached. But the idea was that if the queue is empty the caller could meanwhile flush the result. I forgot why it is not needed that anymore. I may think about it, should not be a relevant difference.

Reply:
I am happy you asked that. Since the current implementation returns as many CUs as available we know the next call would block until new units available=> the CPU has time to flush. I added flush after every call. This way it looks like the flush would take longer (more often only few files flushed), but the overall build time gets even faster, as the latency reduces:
![image](https://github.com/user-attachments/assets/61a60231-11b2-499e-9efc-f0978b4b7936)